### PR TITLE
Add Cosmos3 sound generation

### PR DESCRIPTION
**Pull request by MaciejBalaNV**
## Purpose
This PR is a follow-up to https://github.com/vllm-project/vllm-omni/pull/3454 and adds sound generation to Cosmos3. Already  partially reviewed in https://github.com/MaciejBalaNV/vllm-omni/pull/1

Cosmos3 model is available under https://huggingface.co/nvidia/Cosmos3-Nano (and more variants)
### Unit tests

cd tests; python -m pytest -v -m "core_model and cpu"

Added 23 new test cases for the new model integration and pipeline unit tests.
### Serving tests

Host server with

`vllm serve nvidia/Cosmos3-Nano --omni`

Run a request with
`
curl -sS -X POST http://localhost:8000/v1/videos/sync     -H "Accept: video/mp4"     -F "prompt=A low-angle tracking shot follows a man riding a vintage black motorcycle across a lush green grassy yard. Sunlight filters through overhead trees, casting dappled shadows across the vibrating chrome exhaust and the rider's leather jacket. He kicks up small blades of grass as he maneuvers the bike. He gradually decelerates, the front fork compressing slightly as he brakes to a smooth halt beside another individual standing in the shade. The camera settles into a medium two-shot, capturing the rider lifting his visor to speak, his face framed by a matte helmet. The video is 8 seconds long and is of 24 FPS. This video is of 1280x720 resolution. Audio description: The rhythmic, mechanical chugging of a four-stroke motorcycle engine dominates the foreground, characterized by a throaty, guttural timbre. Periodic high-pitched revs punctuate the steady idle as the throttle is twisted. The sound of tires crunching softly over dry grass and twigs provides a textured background layer. As the vehicle slows, the engine note drops to a low-frequency rumble before clicking into neutral. A muffled, mid-range male voice begins speaking, accompanied by the metallic clink of a helmet visor snapping upward and the faint chirping of distant birds in an open-air environment."     -F "negative_prompt=blurry, distorted, low quality"     -F "size=1280x720"     -F "num_frames=193"     -F "fps=24"     -F "num_inference_steps=35"     -F "guidance_scale=4.0"     -F "seed=42"     -F "generate_sound=true"     -F "sound_duration=8"  -F 'extra_params={"use_resolution_template":false,"use_duration_template":false}'   -o cosmos3_t2v_with_sound.mp4 `
 
## Test Result
The unit tests pass, including all of the new Cosmos unit tests.

===== 3173 passed, 14 skipped, 1511 deselected, 54 warnings in 327.61s (0:05:27) ===
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 1]
```diff
@@ -194,7 +232,19 @@ def post_process_func(
...
+        if isinstance(audio, torch.Tensor):
+            audio = audio.detach().cpu()
+        result = {
+            "video": processed_video,
+            "audio": audio,
+            "fps": _resolve_output_fps(sampling_params),
+        }
+        if audio_sample_rate is not None:
+            result["audio_sample_rate"] = int(audio_sample_rate)
+        return result
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:247]: The post-process function now returns a dict `{video, audio, fps, audio_sample_rate}` when sound is enabled, but still returns a bare tensor/ndarray for video-only paths. This asymmetry may surprise downstream consumers. Consider always returning a dict (even for video-only) or documenting the contract clearly so callers know to check `isinstance(result, dict)`.

File: vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py [hunk 2]
```diff
@@ -138,6 +183,30 @@ def compute_mrope_position_ids_vision(
...
+def compute_mrope_position_ids_sound(
+    grid_t: int,
+    temporal_offset: int | float,
+    sound_latent_fps: float,
+    base_fps: float = 24.0,
+    temporal_compression_factor_sound: int = 1,
+    enable_fps_modulation: bool = True,
+    base_temporal_compression_factor: int | None = None,
+) -> tuple[torch.Tensor, int | float]:
+    """Generate mRoPE IDs for sound tokens as a (T, 1, 1) grid."""
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py:209]: Nit: `del base_temporal_compression_factor` immediately discards this parameter. Consider removing it from the signature entirely (or renaming to `**_kwargs` if part of a shared interface).

File: vllm_omni/diffusion/models/cosmos3/sound_tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+DEFAULT_SOUND_DIM = 64
+DEFAULT_SOUND_HOP_SIZE = 1920
+DEFAULT_SOUND_LATENT_FPS = DEFAULT_SOUND_SAMPLE_RATE / DEFAULT_SOUND_HOP_SIZE
+DEFAULT_SOUND_NORMALIZE_LATENTS = False
+DEFAULT_SOUND_NORMALIZATION_TYPE = "none"
+DEFAULT_SOUND_TANH_INPUT_SCALE = 1.5
+DEFAULT_SOUND_TANH_OUTPUT_SCALE = 3.5
+DEFAULT_SOUND_TANH_CLAMP = 0.995
+SOUND_TOKENIZER_COMPONENT_NAME = "sound_tokenizer"
+SOUND_TOKENIZER_CHECKPOINT_NAME = "diffusion_pytorch_model.safetensors"
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/cosmos3/sound_tokenizer.py:34]: This helper `_config_get` is identical to the one in `audio_tokenizer/avae.py:52`. Worth consolidating into a shared utility or importing from one place to avoid drift.

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 4]
```diff
@@ -946,21 +1111,82 @@ def _cfg_active_at(t: torch.Tensor) -> bool:
...
+
+        def _assign_step_out(step_out: torch.Tensor | tuple[torch.Tensor, ...]) -> None:
+            nonlocal latents, sound_latents
+            if sound_latents is None:
+                assert isinstance(step_out, torch.Tensor)
+                latents = step_out
+                return
+            if not isinstance(step_out, tuple):
+                raise ValueError("Cosmos3 multimodal diffusion step returned a non-tuple result.")
+            latents = step_out[0]
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:1189]: When `sound_latents` is provided, both video and sound noise predictions are packed into a single tensor and stepped together through the scheduler. This works for Euler-style schedulers where the step is stateless per-timestep, but worth confirming the approach is correct for any scheduler that might maintain internal state across steps (e.g. multistep solvers). If the scheduler is always stateless per-step here, a brief comment noting that assumption would help future readers.

**Comment by hsliuustc0106** [comment]
Thanks for adding sound generation to Cosmos3 — the code is well-structured with thorough test coverage (23 new tests).

One thing missing from the PR description: the model-addition checklist requires profiling data and a baseline comparison (latency, VRAM, output quality vs the upstream/nvidia implementation). For a new capability like this, even a single-row table with bs=1 wall time and peak VRAM measured against the official Cosmos3 repo would be helpful to confirm no regressions.

This isn't blocking given the solid unit test results, but it would strengthen confidence for merge.

**Review by hsliuustc0106 — APPROVED** [review]

**Comment by lishunyang12** [comment]
Let's update the recipe.

**Comment by david6666666** [comment]
Tested end-to-end against the official `nvidia/Cosmos3-Nano` examples.

- **Unit tests:** `tests/diffusion/models/cosmos3/` all pass (32 incl. the sound tokenizer + transformer tests).
- **T2V+sound / I2V+sound** via `/v1/videos/sync` with `generate_sound=true`, `sound_duration=7.875` (official params, 1280x720 / 189f / 35 steps): both return video **+ AAC 48 kHz stereo** audio muxed in, matching the structure of the official `example_t2vs_output.mp4` / `example_i2vs_output.mp4`. The AVAE sound tokenizer loads cleanly (sr=48000, 2ch).

| case | output | official ref | match |
|---|---|---|---|
| T2VS | 1280x720/189f + aac 48000Hz 2ch | 1280x720/189f + aac 48000Hz 2ch | Yes  |
| I2VS | 1280x720/189f + aac 48000Hz 2ch | 1280x720/189f + aac 48000Hz 2ch | Yes |

**Review by lishunyang12 — APPROVED** [review]

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py (hunk 1 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:247]: It seems to be consistent with [LTX-2.3](https://github.com/vllm-project/vllm-omni/blob/main/vllm_omni/diffusion/models/ltx2/pipeline_ltx2_3.py#L100) and [Wan2.2](https://github.com/vllm-project/vllm-omni/blob/main/vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2_s2v.py#L241).

File: vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py [hunk 5]
```diff
@@ -859,9 +928,25 @@ def __init__(
...
+        if sound_dim_value is None:
+            sound_dim_value = _od_config_get(od_config, "io_channels", None)
+        if sound_dim_value is None:
+            sound_dim_value = _od_config_get(od_config, "vocoder_input_dim", None)
+        if sound_dim_value is None:
+            sound_dim_value = _od_config_get(od_config, "latent_ch", None)
+        self.sound_gen = _as_bool(sound_gen_value) if sound_gen_value is not None else sound_dim_value is not None
+        from .sound_tokenizer import get_sound_dim, get_sound_latent_fps
+
+        self.sound_dim = int(sound_dim_value if sound_dim_value is not None else get_sound_dim(od_config))
```
[the remark this answers is not in the document]
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py]: Changed so that `sound_dim` and `sound_latent_fps` are arguments passed to transformer from initialized sound tokenizer

File: vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py (hunk 2 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py:209]: deleted

File: vllm_omni/diffusion/models/cosmos3/sound_tokenizer.py (hunk 3 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/sound_tokenizer.py:34]: I think they serve different purposes. `avae._config_get` finds the value of the first present key in a flat dict. `sound_tokenizer._config_get` is used by `_config_path_get` to recursively descend into the config tree looking for a value. They differ in signatures and semantics, so I'd keep them separate.

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py (hunk 4 above)
Comment (bastefaniak) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:1189]: Comment was added around `Cosmos3OmniDiffusersPipeline.diffuse._pack_joint`

**Comment by lishunyang12** [comment]
DCO failed. Let's solve it.

**Comment by bastefaniak** [comment]
> Thanks for adding sound generation to Cosmos3 — the code is well-structured with thorough test coverage (23 new tests).
> 
> One thing missing from the PR description: the model-addition checklist requires profiling data and a baseline comparison (latency, VRAM, output quality vs the upstream/nvidia implementation). For a new capability like this, even a single-row table with bs=1 wall time and peak VRAM measured against the official Cosmos3 repo would be helpful to confirm no regressions.
> 
> This isn't blocking given the solid unit test results, but it would strengthen confidence for merge.

I tested prompt from MR description against [cosmos-framework](https://github.com/NVIDIA/cosmos-framework) (official repository with pytorch inference code), on RTX 6000 Pro Blackwell, both used around 48GB of memory, inference took around 420s, and quality was comparable

**Comment by lishunyang12** [comment]
Let's also add recipe instructions for cosmos-super.

**Comment by lishunyang12** [comment]
> > Thanks for adding sound generation to Cosmos3 — the code is well-structured with thorough test coverage (23 new tests).
> > One thing missing from the PR description: the model-addition checklist requires profiling data and a baseline comparison (latency, VRAM, output quality vs the upstream/nvidia implementation). For a new capability like this, even a single-row table with bs=1 wall time and peak VRAM measured against the official Cosmos3 repo would be helpful to confirm no regressions.
> > This isn't blocking given the solid unit test results, but it would strengthen confidence for merge.
> 
> I tested prompt from MR description against [cosmos-framework](https://github.com/NVIDIA/cosmos-framework) (official repository with pytorch inference code), on RTX 6000 Pro Blackwell, both used around 48GB of memory, inference took around 420s, and quality was comparable

Will be running against official repo on B300 for dc usage. I assume we have one more MR for action generation right? We will launch new release towards end of this week and 0.22.0 version can have full support.

**Comment by lishunyang12** [comment]
**Tested e2e on B300 — all HTTP 200**

**Nano (sound):** T2VS 120.7s · I2VS 123.4s

**Super (64B, 2× B300):** T2I 6s · T2V 197s · I2V 200s · T2VS 198s · I2VS 201s

Outputs match the official examples (1280×720/189f; AAC 48 kHz stereo for sound cases).
